File: packages/swift/Sources/CatalystSwift/Clients/EpicleseClient.swift
// Licensed under the MIT License
//
// Copyright (c) 2025 Natsune Mochizuki

/// Client for Epiclese (World/Author metadata) API endpoints.
public final class EpicleseClient: Sendable {
  private let client: CatalystSwift

  init(client: CatalystSwift) {
    self.client = client
  }

  public func getAuthors(q: String? = nil, platform: String? = nil) async throws
    -> EpicleseAuthorsResult
  {
    return try await client.request(EpicleseEndpoint.getAuthors(q: q, platform: platform))
  }

  public func createAuthor(
    platformIdentifier: String, platform: String, name: String, id: String? = nil
  ) async throws -> EpicleseAuthor {
    let response: EpicleseAuthorWrapper = try await client.request(
      EpicleseEndpoint.createAuthor(
        data: EpicleseCreateAuthorRequest(
          id: id, platformIdentifier: platformIdentifier, platform: platform, name: name)))
    return response.author
  }

  public func getPlatforms() async throws -> [EpiclesePlatform] {
    let response: EpiclesePlatformsWrapper = try await client.request(EpicleseEndpoint.getPlatforms)
    return response.platforms
  }

  public func getPlatform(id: String) async throws -> EpiclesePlatform {
    let response: EpiclesePlatformWrapper = try await client.request(
      EpicleseEndpoint.getPlatform(id: id))
    return response.platform
  }

  public func getStatusMetadata(statusId: String) async throws -> [String: EpicleseStatusMetadataTag]
  {
    return try await client.request(EpicleseEndpoint.getStatusMetadata(statusId: statusId))
  }

  public func createStatusMetadata(
    statusId: String, tags: [EpicleseCreateStatusMetadataTagRequest]
  ) async throws -> [String: EpicleseStatusMetadataTag] {
    return try await client.request(
      EpicleseEndpoint.createStatusMetadata(statusId: statusId, data: tags))
  }

  public func getWorlds(q: String? = nil, platform: String? = nil, offset: Int? = nil) async throws
    -> [EpicleseWorld]
  {
    let response: EpicleseWorldsResult = try await client.request(
      EpicleseEndpoint.getWorlds(q: q, platform: platform, offset: offset))
    return response.items
  }

  public func createWorld(
    platformIdentifier: String, platform: String, name: String,
    author: EpicleseCreateWorldAuthorRequest? = nil, id: String? = nil
  ) async throws -> EpicleseWorld? {
    let response: EpicleseWorldWrapper = try await client.request(
      EpicleseEndpoint.createWorld(
        data: EpicleseCreateWorldRequest(
          id: id, platformIdentifier: platformIdentifier, platform: platform, name: name,
          author: author)))
    return response.world
  }

  public func resolveWorld(platform: String, name: String) async throws -> EpicleseWorld? {
    let response: EpicleseWorldWrapper = try await client.request(
      EpicleseEndpoint.resolveWorld(platform: platform, name: name))
    return response.world
  }
}

File: packages/nodejs/src/clients/epicleseClient.ts
import { EpicleseEndpoint } from "../endpoints/epicleseEndpoint.js";
import type { HttpClient } from "./httpClient.js";
import type {
  EpicleseAuthor,
  EpicleseAuthorWrapper,
  EpicleseAuthorsResult,
  EpiclesePlatform,
  EpiclesePlatformsWrapper,
  EpiclesePlatformWrapper,
  EpicleseWorld,
  EpicleseWorldWrapper,
  EpicleseWorldsWrapper,
  EpicleseCreateAuthorRequest,
  EpicleseCreateWorldRequest,
  EpicleseCreateStatusMetadataTagRequest,
  EpicleseStatusMetadata,
} from "../types/epiclese.js";

/** Client for Epiclese (world/author metadata) API endpoints. */
export class EpicleseClient {
  constructor(private readonly http: HttpClient) {}

  getAuthors(q?: string, platform?: string): Promise<EpicleseAuthorsResult> {
    return this.http.request(EpicleseEndpoint.getAuthors(q, platform));
  }

  async createAuthor(data: EpicleseCreateAuthorRequest): Promise<EpicleseAuthor> {
    const response = await this.http.request<EpicleseAuthorWrapper>(
      EpicleseEndpoint.createAuthor(data),
    );
    return response.author;
  }

  async getPlatforms(): Promise<EpiclesePlatform[]> {
    const response = await this.http.request<EpiclesePlatformsWrapper>(
      EpicleseEndpoint.getPlatforms(),
    );
    return response.platforms;
  }

  async getPlatform(id: string): Promise<EpiclesePlatform> {
    const response = await this.http.request<EpiclesePlatformWrapper>(
      EpicleseEndpoint.getPlatform(id),
    );
    return response.platform;
  }

  getStatusMetadata(statusId: string): Promise<EpicleseStatusMetadata> {
    return this.http.request(EpicleseEndpoint.getStatusMetadata(statusId));
  }

  createStatusMetadata(
    statusId: string,
    tags: EpicleseCreateStatusMetadataTagRequest[],
  ): Promise<EpicleseStatusMetadata> {
    return this.http.request(
      EpicleseEndpoint.createStatusMetadata(statusId, tags),
    );
  }

  async getWorlds(q?: string, platform?: string, offset?: number): Promise<EpicleseWorld[]> {
    const response = await this.http.request<EpicleseWorldsWrapper>(
      EpicleseEndpoint.getWorlds(q, platform, offset),
    );
    return response.items;
  }

  async createWorld(data: EpicleseCreateWorldRequest): Promise<EpicleseWorld | null> {
    const response = await this.http.request<EpicleseWorldWrapper>(
      EpicleseEndpoint.createWorld(data),
    );
    return response.world;
  }

  async resolveWorld(platform: string, name: string): Promise<EpicleseWorld | null> {
    const response = await this.http.request<EpicleseWorldWrapper>(
      EpicleseEndpoint.resolveWorld(platform, name),
    );
    return response.world;
  }
}
